lewis/adapters/modbus.py

```diff
@@ -0,0 +1,450 @@
+# -*- coding: utf-8 -*-
+# *********************************************************************
+# lewis - a library for creating hardware device simulators
+# Copyright (C) 2016 European Spallation Source ERIC
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# *********************************************************************
+# References Used:
+#   http://www.modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf
+#   http://www.modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf
+#   https://github.com/sourceperl/pyModbusTCP
+#   https://github.com/bashwork/pymodbus
+# *********************************************************************
+
+
+import socket
+import asyncore
+import struct
+import inspect
+
+from copy import deepcopy
+from math import ceil
+from collections import OrderedDict
+
+from lewis.core.statemachine import StateMachine
+from lewis.adapters import Adapter
+
+
+class MBFC(object):
+    """Modbus standard function codes"""
+    READ_COILS = 0x01
+    READ_DISCRETE_INPUTS = 0x02
+    READ_HOLDING_REGISTERS = 0x03
+    READ_INPUT_REGISTERS = 0x04
+    WRITE_SINGLE_COIL = 0x05
+    WRITE_SINGLE_REGISTER = 0x06
+    WRITE_MULTIPLE_COILS = 0x0F
+    WRITE_MULTIPLE_REGISTERS = 0x10
+    # MODBUS_ENCAPSULATED_INTERFACE = 0x2B
+
+    @classmethod
+    def is_valid(cls, code):
+        # Adapted from: http://stackoverflow.com/a/4241225/3827434
+        boring = dir(type('dummy', (object,), {}))
+        codes = dict([attr for attr in inspect.getmembers(cls) if attr[0] not in boring])
+        return code in codes.values()
+
+
+class MBEX(object):
+    """Modbus standard exception codes"""
+    NONE = 0x00
+    ILLEGAL_FUNCTION = 0x01
+    DATA_ADDRESS = 0x02
+    DATA_VALUE = 0x03
+    SLAVE_DEVICE_FAILURE = 0x04
+    ACKNOWLEDGE = 0x05
+    SLAVE_DEVICE_BUSY = 0x06
+    MEMORY_PARITY_ERROR = 0x08
+    GATEWAY_PATH_UNAVAILABLE = 0x0A
+    GATEWAY_TARGET_DEVICE_FAILED_TO_RESPOND = 0x0B
+
+
+class ModbusBasicDatabank(object):
+    def __init__(self):
+        self.bits = [False] * 0x10000
+        self.words = [0] * 0x10000
```

Comment:
Sounds like a good idea, I'll have a look. Thanks!

This basic databank is pretty much copied straight from pyModbusTCP and just there as a placeholder, it need a lot more work. Especially since it needs to interact with the Interface/Device.